### DISCUSSION TITLE
请教下集群模式如何实现指定的节点能看到运算结果

### DISCUSSION THREAD

**Discussion by mingo0117** [secretflow/secretflow#46] [Q&A]
小白一枚，想请教下：
例如我现在有三个节点(alice,bob,charlies)，各自单独部署。alice和bob是计算方，charlies是结果方，需要**仅charlies能看到alice和bob相加的结果**（alice和bob互相不知道对方的输入，也不能知道最终的结果）。看目前的例子不知道如何实现，望指点!

讲一下我的步骤：
1、创建ray（我目前是希望在单机上模拟集群）
ray start --head --node-ip-address="172.17.0.12" --port="6379" --resources='{"alice": 1}'
ray start --address="172.17.0.12:6379" --resources='{"bob": 1}'
ray start --address="172.17.0.12:6379" --resources='{"charlie": 1}'

2、相加
```

import secretflow as sf
sf.shutdown()
sf.init(address='172.17.0.12:6379')
alice, bob, charlie = sf.PYU('alice'), sf.PYU('bob'), sf.PYU('charlie')
import numpy as np
a = alice(lambda : np.random.rand(2, 5))()
b = bob(lambda : np.random.rand(2, 5))()
##不确定这样写是否代表只有charlie能看到alice+bob
secure_aggr = sf.security.aggregation.SecureAggregator(device=charlie, participants=[alice, bob])
sf.reveal(secure_aggr.sum([a, b], axis=0))
array([[0.99529266, 1.23123932, 1.38018417, 1.0768013 , 1.56233215],
       [0.59137726, 0.88001251, 1.03471756, 1.1807251 , 1.29380035]])

```
我的疑问是：如何证明（或者说演示）alice和bob是看不到这个array结果的，只有charlie能看到？
之前看过矩阵元的实现（https://github.com/LatticeX-Foundation/Rosetta/blob/master/example/tutorials/code/rosetta_demo.py），
三方部署在三个节点（先启动charlie后启动alice和bob），跑同样的代码，但是只有指定方charlie的reveal才能看到结果，且我可以打开log看到协议层是如何实现的。那么如果用secretflow，我需要怎么做才能体现出来呢？

**Comment by zhouaihui** [comment]
这个[issue](https://github.com/secretflow/secretflow/issues/50)应该回答了一部分你的疑问。

**Comment by zhouaihui** [comment]
> ##不确定这样写是否代表只有charlie能看到alice+bob
secure_aggr = sf.security.aggregation.SecureAggregator(device=charlie, participants=[alice, bob])

是的，这行代码表示由charlie聚合alice和bob的数据，charlie知道最终聚合的结果，alice和bob是看不到的。

**Comment by mingo0117** [comment]
嗯。这个我已经理解了，谢谢。issue就是我提的
